File: frontend/mobile/senserx/lib/domain/models/product_details.dart
class ProductDetails {
  final String ean;
  final String title;
  final String description;
  final String upc;
  final String brand;
  final String model;
  final String color;
  final String size;
  final String dimension;
  final String weight;
  final String category;
  final String currency;
  final double lowestRecordedPrice;
  final double highestRecordedPrice;
  final List<String> images;
  final List<Offer> offers;
  final String asin;
  final String elid;

  ProductDetails({
    required this.ean,
    required this.title,
    required this.description,
    required this.upc,
    required this.brand,
    required this.model,
    required this.color,
    required this.size,
    required this.dimension,
    required this.weight,
    required this.category,
    required this.currency,
    required this.lowestRecordedPrice,
    required this.highestRecordedPrice,
    required this.images,
    required this.offers,
    required this.asin,
    required this.elid,
  });

  factory ProductDetails.fromJson(Map<String, dynamic> json) {
    return ProductDetails(
      ean: json['ean'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      upc: json['upc'] ?? '',
      brand: json['brand'] ?? '',
      model: json['model'] ?? '',
      color: json['color'] ?? '',
      size: json['size'] ?? '',
      dimension: json['dimension'] ?? '',
      weight: json['weight'] ?? '',
      category: json['category'] ?? '',
      currency: json['currency'] ?? '',
      lowestRecordedPrice: (json['lowest_recorded_price'] as num?)?.toDouble() ?? 0.0,
      highestRecordedPrice: (json['highest_recorded_price'] as num?)?.toDouble() ?? 0.0,
      images: (json['images'] as List<dynamic>?)?.cast<String>() ?? [],
      offers: (json['offers'] as List<dynamic>?)
          ?.map((offer) => Offer.fromJson(offer))
          .toList() ?? [],
      asin: json['asin'] ?? '',
      elid: json['elid'] ?? '',
    );
  }

  @override
  String toString() {
    return '''
    ProductDetails {
      ean: $ean,
      title: $title,
      description: $description,
      upc: $upc,
      brand: $brand,
      model: $model,
      color: $color,
      size: $size,
      dimension: $dimension,
      weight: $weight,
      category: $category,
      currency: $currency,
      lowestRecordedPrice: $lowestRecordedPrice,
      highestRecordedPrice: $highestRecordedPrice,
      images: $images,
      offers: $offers,
      asin: $asin,
      elid: $elid
    }
    ''';
  }
}

class Offer {
  final String merchant;
  final String domain;
  final String title;
  final String currency;
  final dynamic listPrice;
  final dynamic price;
  final String shipping;
  final String condition;
  final String availability;
  final String link;
  final int updatedT;

  Offer({
    required this.merchant,
    required this.domain,
    required this.title,
    required this.currency,
    required this.listPrice,
    required this.price,
    required this.shipping,
    required this.condition,
    required this.availability,
    required this.link,
    required this.updatedT,
  });

  factory Offer.fromJson(Map<String, dynamic> json) {
    return Offer(
      merchant: json['merchant'] ?? '',
      domain: json['domain'] ?? '',
      title: json['title'] ?? '',
      currency: json['currency'] ?? '',
      listPrice: json['list_price'] ?? '',
      price: json['price'] ?? 0,
      shipping: json['shipping'] ?? '',
      condition: json['condition'] ?? '',
      availability: json['availability'] ?? '',
      link: json['link'] ?? '',
      updatedT: json['updated_t'] ?? 0,
    );
  }

  @override
  String toString() {
    return '''
    Offer {
      merchant: $merchant,
      domain: $domain,
      title: $title,
      currency: $currency,
      listPrice: $listPrice,
      price: $price,
      shipping: $shipping,
      condition: $condition,
      availability: $availability,
      link: $link,
      updatedT: $updatedT
    }
    ''';
  }
}

File: backend/api/src/domain/models/Products/ProductDetails.ts
import {BaseModel, validationSchemaKey} from "../BaseModel";
import { z } from 'zod';
import {Schema} from "redis-om";

/**
 * Validation Schema
 */
export const productDetailsSchema = z.object({
  ean: z.string(),
  title: z.string(),
  description: z.string(),
  upc: z.string(),
  brand: z.string(),
  model: z.string(),
  color: z.string(),
  size: z.string(),
  dimension: z.string(),
  weight: z.string(),
  category: z.string(),
  currency: z.string(),
  lowest_recorded_price: z.string(),
  highest_recorded_price: z.string(),
  images: z.array(z.string())
});

/**
 * Product Details Type Definition
 */
export type ProductDetails = z.infer<typeof productDetailsSchema>;

/**
 * ProductDetails Model
 */
export class ProductDetailsModel extends BaseModel<ProductDetails> {
  [validationSchemaKey] = productDetailsSchema;

  ean: string;
  title: string;
  description: string;
  upc: string;
  brand: string;
  model: string;
  color: string;
  size: string;
  weight: string;
  category: string;
  currency: string;
  dimension: string;
  lowest_recorded_price: string;
  highest_recorded_price: string;
  images: string[];

  constructor(data: any) {
      super();
      this.ean  = data.ean;
      this.title = data.title;
      this.description = data.description;
      this.upc = data.upc;
      this.brand = data.brand;
      this.model = data.model;
      this.color = data.color;
      this.size = data.size;
      this.weight = data.weight;
      this.category = data.category;
      this.currency = data.currency;
      this.lowest_recorded_price = data.lowest_recorded_price;
      this.highest_recorded_price = data.highest_recorded_price;
      this.images = data.images;
      this.dimension = data.dimension;
  }

  static schema () {
    return new Schema(
        this.name,
        {
          upc: { type: 'string', indexed: true },
          ean: { type: 'string' },
          title: { type: 'string' },
          description: { type: 'text' },
          brand: { type: 'string' },
          model: { type: 'string' },
          color: { type: 'string' },
          size: { type: 'string' },
          dimension: { type: 'string' },
          weight: { type: 'string' },
          category: { type: 'string' },
          currency: { type: 'string' },
          lowest_recorded_price: { type: 'string' },
          highest_recorded_price: { type: 'string' },
          images: { type: 'string[]' }
      })
  }
}